pkg/controllers/provisioningv2/rke2/rkecluster/controller.go

```diff
@@ -2,73 +2,85 @@ package rkecluster
-		"rke-cluster-status",
-		h.OnChange)
+	clients.RKE.RKECluster().OnChange(ctx, "rke-cluster", h.OnChange)
 	relatedresource.Watch(ctx, "rke-cluster-trigger", func(namespace, name string, obj runtime.Object) ([]relatedresource.Key, error) {
-		return []relatedresource.Key{{
-			Namespace: namespace,
-			Name:      name,
-		}}, nil
-	}, clients.RKE.RKECluster(), clients.RKE.RKEControlPlane())
+		if capiCluster, ok := obj.(*capi.Cluster); ok && !capiCluster.Spec.Paused {
+			return []relatedresource.Key{{
+				Namespace: namespace,
+				Name:      name,
+			}}, nil
+		}
+		return nil, nil
+	}, clients.RKE.RKECluster(), clients.CAPI.Cluster())
 }
 
-func (h *handler) UpdateSpec(_ string, cluster *v1.RKECluster) (*v1.RKECluster, error) {
+func (h *handler) OnChange(_ string, cluster *v1.RKECluster) (*v1.RKECluster, error) {
 	if cluster == nil {
 		return nil, nil
 	}
 
-	if cluster.Spec.ControlPlaneEndpoint == nil {
+	capiCluster, err := rke2.GetOwnerCAPICluster(cluster, h.capiClusterCache)
+	if apierrors.IsNotFound(err) {
+		logrus.Debugf("[rkecluster] %s/%s: waiting: CAPI cluster does not exist", cluster.Namespace, cluster.Name)
+		h.rkeCluster.EnqueueAfter(cluster.Namespace, cluster.Name, 10*time.Second)
+		return cluster, generic.ErrSkip
+	}
+	if err != nil {
+		logrus.Errorf("[rkecluster] %s/%s: error getting CAPI cluster %v", cluster.Namespace, cluster.Name, err)
+		return cluster, err
+	}
+
+	if capiannotations.IsPaused(capiCluster, cluster) {
+		logrus.Infof("[rkecluster] %s/%s: waiting: CAPI cluster or RKECluster is paused", cluster.Namespace, cluster.Name)
+		return cluster, generic.ErrSkip
+	}
+
+	if cluster.Spec.ControlPlaneEndpoint == nil || !cluster.Spec.ControlPlaneEndpoint.IsValid() {
 		cluster := cluster.DeepCopy()
-		cluster.Spec.ControlPlaneEndpoint = &v1.Endpoint{
+		cluster.Spec.ControlPlaneEndpoint = &capi.APIEndpoint{
 			Host: "localhost",
 			Port: 6443,
 		}
-		return h.clusterClient.Update(cluster)
+		logrus.Debugf("[rkecluster] %s/%s: setting controlplane endpoint", cluster.Namespace, cluster.Name)
+		return h.rkeCluster.Update(cluster)
 	}
 
-	return cluster, nil
-}
-
-func (h *handler) OnChange(rkeCluster *v1.RKECluster, status v1.RKEClusterStatus) (v1.RKEClusterStatus, error) {
-	conditionToUpdate := rke2.Ready
-	if !rkeCluster.DeletionTimestamp.IsZero() {
-		conditionToUpdate = rke2.Removed
-	}
-	cp, err := h.rkeControlPlanes.Get(rkeCluster.Namespace, rkeCluster.Name)
-	if err == nil {
-		conditionToUpdate.SetStatus(&status, conditionToUpdate.GetStatus(cp))
-		conditionToUpdate.Reason(&status, conditionToUpdate.GetReason(cp))
-		conditionToUpdate.Message(&status, conditionToUpdate.GetMessage(cp))
-	} else if !apierrors.IsNotFound(err) {
-		return status, err
+	if len(cluster.Status.Conditions) > 0 || cluster.Status.Ready != true {
+		cluster := cluster.DeepCopy()
+		// the rke2.Ready and rke2.Removed conditions may still be present on the object, remove them if present
+		cluster.Status.Conditions = nil
```

Comment:
how does rancher react with the conditions on the provisioning cluster being missing?

Reply:
This is the infrastructure cluster, not the provisioning cluster, so neither rancher nor CAPI utilize these conditions.